service/handlers.go

```diff
@@ -0,0 +1,437 @@
+		id, err := strconv.Atoi(mux.Vars(r)["id"])
+		if err != nil {
+			s.log.Printf("Error parsing asset ID: %v", err)
+			http.Error(w, "Invalid asset ID", http.StatusBadRequest)
+			return
+		}
+
+		asset, err := s.storage.assetWithRLock(uint(id))
+		if err != nil {
+			s.log.Printf("Error getting asset: %v", err)
+			http.Error(w, "Asset not found", http.StatusNotFound)
+			return
+		}
+		assetBytes, err := json.Marshal(asset)
+		if err != nil {
+			s.log.Printf("Error marshalling asset: %v", err)
+			http.Error(w, "Error marshalling asset", http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.Write(assetBytes)
+	}
+}
+
+func (s *Service) handleGetAssets() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+		typeStr := r.URL.Query().Get("type")
+		assets := s.storage.getAssets(assetType(typeStr))
+		err := json.NewEncoder(w).Encode(assets)
+		if err != nil {
+			// Before the error occurs, we may have already sent
+			// some payload to the client. If that's the case,
+			// response headers are already set (i.e 200 OK).
+			// Clients can detect failure by not receiving a
+			// zero-length chunk in chunked transfer encoding.
+			// https://stackoverflow.com/questions/17203379/response-sent-in-chunked-transfer-encoding-and-indicating-errors-happening-after
+			s.log.Printf("Error sending assets: %v", err)
+			http.Error(w, "Error marshalling assets", http.StatusInternalServerError)
+			return
+		}
+
+	}
+}
+
+func (s *Service) handleCreateAsset() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var a asset
+		err := json.NewDecoder(r.Body).Decode(&a)
+		if err != nil {
+			s.log.Printf("Error decoding asset: %v", err)
+			http.Error(w, "Error decoding asset", http.StatusBadRequest)
+			return
+		}
+		err = a.validate()
+		if err != nil {
+			msg := fmt.Sprintf("Error validating asset: %v", err)
+			s.log.Print(msg)
+			http.Error(w, msg, http.StatusBadRequest)
+			return
+		}
+		s.storage.addAsset(&a)
+		s.metrics.Add("assets", 1)
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+func (s *Service) handleUpdateAsset() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		id, err := strconv.Atoi(mux.Vars(r)["id"])
+		if err != nil {
+			s.log.Printf("Error parsing asset ID: %v", err)
+			http.Error(w, "Invalid asset ID", http.StatusBadRequest)
+			return
+		}
+
+		var a asset
+		err = json.NewDecoder(r.Body).Decode(&a)
+		if err != nil {
+			s.log.Printf("Error decoding asset: %v", err)
+			http.Error(w, "Error decoding asset", http.StatusBadRequest)
+			return
+		}
+		a.ID = uint(id)
+
+		err = s.storage.updateAsset(&a)
+		if err != nil {
+			s.log.Printf("Error updating asset: %v", err)
+			http.Error(w, "Asset ID not found", http.StatusNotFound)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+func (s *Service) handleDeleteAsset() http.HandlerFunc {
```

Comment:
why do you return `http.HandlerFunc` instead of implementing handler in `handleDeleteAsset` directly?